runtime/intern.c

```diff
@@ -722,66 +752,123 @@ static void caml_parse_header(struct caml_intern_state* s,
   case Intext_magic_number_big:
 #ifdef ARCH_SIXTYFOUR
     h->header_len = 32;
+    h->compressed = 0;
     read32u(s);
-    h->data_len = read64u(s);
+    h->data_len = h->uncompressed_data_len = read64u(s);
     h->num_objects = read64u(s);
     h->whsize = read64u(s);
 #else
-    errmsg[sizeof(errmsg) - 1] = 0;
-    snprintf(errmsg, sizeof(errmsg) - 1,
-             "%s: object too large to be read back on a 32-bit platform",
-             fun_name);
-    caml_failwith(errmsg);
+    intern_failwith2
+      (fun_name, "object too large to be read back on a 32-bit platform");
 #endif
     break;
+  case Intext_magic_number_compressed:
+    h->header_len = read8u(s) & 0x3F;
+    h->compressed = 1;
+    int overflow = 0;
+    overflow |= readvlq(s, &h->data_len);
+    overflow |= readvlq(s, &h->uncompressed_data_len);
+    overflow |= readvlq(s, &h->num_objects);
+#ifdef ARCH_SIXTYFOUR
+    (void) readvlq(s, NULL);
+    overflow |= readvlq(s, &h->whsize);
+#else
+    overflow |= readvlq(s, &h->whsize);
+    (void) readvlq(s, NULL);
+#endif
+    if (overflow) {
+      intern_failwith2
+        (fun_name, "object too large to be read back on this platform");
+    }
+    break;
   default:
-    errmsg[sizeof(errmsg) - 1] = 0;
-    snprintf(errmsg, sizeof(errmsg) - 1,
-             "%s: bad object",
-             fun_name);
-    caml_failwith(errmsg);
+    intern_failwith2(fun_name, "bad object");
   }
 }
 
+/* Decompress the input if needed.
+   Must be called after [intern_init].
+   Should preferably be called before [intern_alloc_storage]
+   (so that the memory block for the compressed input can be freed
+    before more memory is allocated). */
+
+static void intern_decompress_input(struct caml_intern_state * s,
+                                    char * fun_name,
+                                    struct marshal_header * h)
+{
+  s->compressed = h->compressed;
+  if (! h->compressed) return;
+#ifdef HAS_ZSTD
+  unsigned char * blk = caml_stat_alloc_noexc(h->uncompressed_data_len);
+  if (blk == NULL) {
+    intern_cleanup(s);
+    caml_raise_out_of_memory();
+  }
+  size_t res =
+    ZSTD_decompress(blk, h->uncompressed_data_len, s->intern_src, h->data_len);
+  if (res != h->uncompressed_data_len) {
+    caml_stat_free(blk);
+    intern_cleanup(s);
+    intern_failwith2(fun_name, "decompression error");
```

Comment:
There is `ZSTD_isError` , `ZSTD_getErrorCode` and `ZSTD_getErrorName` that can be used to show more details on what went wrong (see `zstd_errors.h`).
In particular since the rest of the code handles out of memory it'd be nice if ZSTD_error_memory_allocation was handled (or just a generic function that raises an exception using some prefix + the error from zstd?). Printing a more detailed error message might help if the stream becomes truncated/corrupted.

Reply:
I tried to use `ZSTD_getErrorName` to get better error messages, see https://github.com/ocaml/ocaml/commit/4751652ba29acbee1fd11031dba6be0000c9a8b4 .   Then I tried to trigger a decompression error by fuzzing bytes in marshaled data, and could not: either the fuzzing just changes the data after unmarshaling, or unmarshaling crashes.  This suggests that ZTSD, by default, does not CRC the data.  In the end, I'd rather keep error reporting minimal.

Reply:
If you still have it, could you send me the fuzzed data that causes unmarshaling to crash via e-mail?
(you can see my e-mail in one of the commits, e.g. .https://github.com/edwintorok/ocaml/commit/d6e1755372e0ab5083dd0bda45d15c10bc44343b)

'zstd' is used in many other programs and you might've discovered a security bug (best to discuss the details of that via e-mail though).

Reply:
Oh, I'm 100% sure ZSTD did not crash: it expanded the fuzzed compressed data, resulting in ill-formed marshaled data, and then the unmarshaler crashed.

Reply:
Ok, nothing to worry about then, I don't think OCaml would support reading in marshalled data from untrusted sources, and if zstd itself doesn't crash no need to worry about other projects being affected. Lets leave the error handling as it is then.